### DISCUSSION TITLE
Disable auto linking in part of Markdown file

### DISCUSSION THREAD

**Discussion by itavero** [doxygen/doxygen#10992] [Q&A]
We have several Markdown (`.md`) files that contain Mermaid diagrams.
I managed to get this working by using an filter script similar to [this one](https://github.com/tahyx/doxygen-mermaid/blob/feature/md_mermaid_example/doxygen_filter_mermaid.sh).

However, if a diagram contains a name that Doxygen can link, it will mess up the mermaid code by adding an `<a>` tag in it.
In general I like the idea of auto linking, but to my surprise I could not find a way to tell Doxygen to not perform auto linking in a certain section.
How can this be achieved?

Adding `%` doesn't work that well, as that would break rendering in other viewers, and symbols/names may change/be added after a diagram has been made.

**Comment by itavero** [comment]
For now, as I workaround, my filter script performs a Base64 encoding of the Mermaid.js code.
In the footer HTML I transform this back into regular text using javascript before calling `mermaid.initialize`.

Something like:
```html
<script>
  // To prevent Doxygen from messing up the formatting of the Mermaid.js input,
  // we use a Base64 encoded version of the input in the pre.mermaid-b64.
  // Before calling mermaid.initialize(), we decode the content of all pre.mermaid-b64 elements
  // and change the class to pre.mermaid.
  document.querySelectorAll('pre.mermaid-b64').forEach((element) => {
    const decoded = atob(element.textContent);
    element.textContent = decoded;
    element.classList.remove('mermaid-b64');
    element.classList.add('mermaid');
  });
  mermaid.initialize({
    startOnLoad: true,
    theme: 'base',
  });
</script>
```

Of course, if I could just tell Doxygen to not bother with linking in the mermaid `<pre>` tag, that would still be preferred.

**Comment by albert-github** [comment]
There are no commands, to my knowledge, that would accomplish this.
Recently a feature was added to doxygen (issue #10959 pull request #10966) that added a special comment to doxygen (`<!--!...-->` ) that is ignored in other rederers but that doxyge sees and outputs the `...` part.

This would lead to that e.g.
```
<!--!%-->Cls
```
should be rendered as `%Cls` in doxygen and should be seen by other  renderers as `Cls`.
This fix has one flaw thought which is corrected in the proposed pull request #10933

Probably still not the  easiest way but no extra software is needed.

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub.

**Comment by richarddeurloo** [comment]
@itavero Would you be willing to share your script? I have struggling with the same issue. And I don't think the special comment to doxygen above would be best way to go.

I'm also using a script based on the script from tahyx that you linked in your initial post. In particular, how does your "filter script performs a Base64 encoding of the Mermaid.js code"?

**Comment by albert-github** [comment]
@richarddeurloo 
> And I don't think the special comment to doxygen above would be best way to go.

- Did you try the `<!--! ... -->` syntax?
- which problems did you encounter

**Reply by richarddeurloo** [reply]
No, I didn't try. I may be misunderstanding the feature, but I don't think it will solve the issue. As stated the content of the special HTML comment will get ignored by regular Markdown processors but Doxygen will see it. But my issue (and I believe @itavero as well) is different.

We have a markdown file that contains a mermaid diagram. For example:
~~~
# Flow Diagram

```mermaid
flowchart TD
    ClassName1 --> ClassName2 
```
~~~
When viewing this markdown file in GitHub or an IDE (e.g. VS Code), the diagram is automatically rendered. But to ensure it is rendered correctly in HTML pages generated by Doxygen, some extra steps are needed.

AFAIK there is no native support for Mermaid diagrams (yet) in Doxygen, like for PlantUML. So I use the script from [tahyx](https://github.com/tahyx/doxygen-mermaid/blob/feature/md_mermaid_example/doxygen_filter_mermaid.sh) to preprocess the markdown file in Doxygen using the `FILTER_PATTERNS` mechanism. This script replaces the code block with:

~~~
<pre class="mermaid">
--
flowchart TD
    ClassName1 --> ClassName2 
</pre>
~~~

The problem is that Doxygen will recognise the classnames and try to autolinks text, resulting in:

~~~
<pre class="mermaid">
--
flowchart TD
    <a class="el" href="[linkname](url)">ClassName1</a> --> <a class="el" href="[linkname](url)">ClassName2</a>
</pre>
~~~
 
This now contains illegal syntax that cannot be rendered by the Mermaid renderers.

So I don't think the special HTML comment will work in this case, at least not in a easy-to-use way. Ideally, we could locally disable the autolinking or have native Mermaid support (that would effectively do the same).

But in the short term I'd like to understand @itavero solution and solve the problem I'm having now.

**Comment by albert-github** [comment]
Indeed there is no support for Mermaid diagrams, to be able to do it an approach similar to plantuml could be used but this at least requires some easy to use, free, well maintained Mermaid renderer software. When at least these preconditions are satisfied someone still has to implement it ...

**Reply by richarddeurloo** [reply]
I was planning to file a ticket with a request for Mermaid support if there isn't one yet already (didn't check yet). And potentially work on a PR. But I currently do not have the bandwidth to do so.

Just FYI the Mermaid renderer is well maintained: https://github.com/mermaid-js/mermaid. But it's a little different from for example PlantUML, because for HTLM at least it renders the diagrams dynamically using JavaScript: https://mermaid.js.org/intro/#deploying-mermaid. For PDF this would require a different approach. There is a CLI version as well that may serve this purpose: https://github.com/mermaid-js/mermaid-cli.

**Reply by albert-github** [reply]
There is an issue related to this #11445
Problem I see here is that it looks like it just `js` code, question would be how to integrate this into other formats than HTML, in other words how to generates png / svg images out of it.

**Reply by richarddeurloo** [reply]
The CLI version of Mermaid is able to do that. See: [Transform a markdown file with mermaid diagrams](https://github.com/mermaid-js/mermaid-cli?tab=readme-ov-file#transform-a-markdown-file-with-mermaid-diagrams). But I have not given it much thought or looked at the Doxygen PlantUML approach. I check out the issue and comment there.

**Reply by albert-github** [reply]
I didn't study it either but the `mmdc` approach look interesting.
